source/agora/node/API.d

```diff
@@ -198,4 +200,75 @@ public interface API
     ***************************************************************************/
 
     public Hash[] getMerklePath (ulong block_height, Hash hash);
+
+
+    /***************************************************************************
+
+        Get UTXOs from the UTXO set
+
+        API:
+            GET /get_utxos
+
+        Params:
+            address = used to identify the owner of UTXOs
+            type = transaction type of UTXOs to find
+            min_amount_value = mininum amount of UTXOs to find
+
+        Returns:
+            the associative array for UTXOs
+
+    ***************************************************************************/
+
+    public UTXOSetValue[Hash] getUTXOs (PublicKey address, TxType type,
+        ulong min_amount_value);
+
+
+    /***************************************************************************
+
+        Create an enrollment data for enrollment process
+
+        Params:
+            frozen_utxo_hash = the hash of a frozen UTXO used to identify
+                        a validator or an enrollment data
+
+        Returns:
+            the Enrollment object created
+
+    ***************************************************************************/
+
+    public Enrollment createEnrollment (Hash frozen_utxo_hash);
+
+
+    /***************************************************************************
+
+        Enroll as a validator
+
+        API:
+            PUT /enroll_validator
+
+        Params:
+            enroll = the Enrollment object, the information about an validator
+
+    ***************************************************************************/
+
+    public void enrollValidator (Enrollment enroll);
+
+
+    /***************************************************************************
+
+        Check if an enrollment data exists in the validator set
+
+        Params:
+            enroll_hash = key for an enrollment data which is hash of frozen UTXO
+
+        API:
+            GET /has_enrollment
+
+        Returns:
+            true if the validator set has the enrollment data
+
+    ***************************************************************************/
+
+    @method(HTTPMethod.GET)
+    public bool hasEnrollment (Hash enroll_hash);
```

Comment:
What is the use-case for this API? I see that it's used in the unittest, but other than the tests why do we need to query another node if it contains an enrollment hash?

Reply:
I follow the way the code about transaction has implemented. It has the hasTransactionHash function so I add the API similar to that. Its usage is just for unittest now.

Reply:
But in that case, `NetworkClient.sendTransaction` calls `hasTransactionHash` before calling `putTransaction`, to avoid sending the same transaction to a node which already contains it. That was its use-case.

Then the same should be done for the enrollment, right? `NetworkManager.sendEnrollment` should first check `hasEnrollment` before calling `enrollValidator` to avoid spamming the network.

Reply:
That's right. I have missed it. Thanks.

Reply:
It is already checked in receiveEnrollment function, actually in the addEnrollment function in the receiveEnrollment function. The sendTransaction you mentioned is not checking other node's state also.

Reply:
But if it's checked in `receiveEnrollment` it means the structure was already sent over the wire. The point is to not send data that doesn't have to be sent.

Reply:
I got it. I will be checking the code more.

Reply:
I will add the hasEnrollment function to the sendEnrollement function of the **NetworkClient** class like Transaction.